### DISCUSSION TITLE
How to use SSG with API Route

### DISCUSSION THREAD

**Discussion by mrvectopus** [shuding/nextra#2747] [Q&A] [answered]
**Hello! 🥲** 

I connected the database and configured the API Route, it looks like this:

_/pages/api/sponsors.ts_

`[{"donate":"slowly reward","group":"temporise inside","group_slug":"out deceivingly","id":"rec_cneaqp2k1n2tltosuvo0","message":"angelic woefully","name":"inquisitively gah","username":"over childhood","xata":{"createdAt":"2024-02-26T15:25:24.894Z","updatedAt":"2024-02-26T15:25:24.894Z","version":0}},{"donate":"supposing whether","group":"ugh crab","group_slug":"aha clearly","id":"rec_cneaqp2k1n2tltosuvog","message":"upwardly uh-huh","name":"register limp","username":"quizzically stockpile","xata":{"createdAt":"2024-02-26T15:25:24.894Z","updatedAt":"2024-02-26T15:25:24.894Z","version":0}}]
`

Now I need to show data `{username}` and `{donate}` from the database on the page `index.mdx` with server-side generation.

Like this:
```
<div>
<h2>Hello, {username}</h2>
<p>You have {donate} donates</p>
</div>

```
I found this method: https://nextra.site/docs/guide/ssg, but this code is not work when I change fetch url to my local API Route (http://localhost:3000/pages/api/sponsors).

Next, I found this method: https://nextjs.org/docs/pages/building-your-application/data-fetching/client-side
I'm create a component with this code and change url to my `/pages/api/sponsors`, this is work, but this method is not a server-side.

I was try this method too: https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props#using-getstaticprops-to-fetch-data-from-a-cms

I'm created in `/pages` a `test.tsx` page and add this code, this is work like SSG, but the page is blank, I mean page show only text, data from `http://localhost:3000/api/sponsors` without a Sidebar, Header and etc... it's not cool.

**Please, help me how I can do it :(**

**Comment by caffeine-addictt** [comment]
Can you provide a link to your repo or give some code snippets to show what you have tried that didn't work?

**Reply by mrvectopus** [reply]
## API Router

This code I'm use to create a API Router `/api/getSponsors.ts` from database:

```
import { getXataClient } from '../../src/xata';

const xata = getXataClient();

const handler = async (req, res) => {
  // getMany or getAll method can be used to create records in database
  try {
    const data = await xata.db.getSponsors.getAll();
    res.json(data);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

export default handler;
```

And this is result on page `http://localhost:3000/api/getSponsors`:

`
[{"donate":"slowly reward","group":"temporise inside","group_slug":"out deceivingly","id":"rec_cneaqp2k1n2tltosuvo0","message":"angelic woefully","name":"inquisitively gah","username":"over childhood","xata":{"createdAt":"2024-02-26T15:25:24.894Z","updatedAt":"2024-02-26T15:25:24.894Z","version":0}},{"donate":"supposing whether","group":"ugh crab","group_slug":"aha clearly","id":"rec_cneaqp2k1n2tltosuvog","message":"upwardly uh-huh","name":"register limp","username":"quizzically stockpile","xata":{"createdAt":"2024-02-26T15:25:24.894Z","updatedAt":"2024-02-26T15:25:24.894Z","version":0}}]
`

## Method 1
https://nextra.site/docs/guide/ssg

I'm add this code to my `index.mdx` page:

```
import { useData } from 'nextra/data'

export const getStaticProps = async ({ params }) => {
  const res = await fetch(`http://localhost:3000/api/getSponsors`);
  const data = await res.json();

  return {
    props: {
      ssg: {
        donate: data.donate || null,
        username: data.username || null,
      },
      revalidate: 60
    }
  };
};

export const Stars = () => {
  const { donate, username } = useData()
  return (
      <span className="text-blue-600">This is donate {donate}, from {username}</span>
  );
};

# Hello!

Page for the test...
```

This is not work, I'm view only my text "Hello! Page for the test...", but when for the test I'm use code from example, it's work:

```
import { useData } from 'nextra/data'
 
export const getStaticProps = ({ params }) => {
  return fetch(`https://api.github.com/repos/shuding/nextra`)
    .then(res => res.json())
    .then(repo => ({
      props: {
        // We add an `ssg` field to the page props,
        // which will be provided to the Nextra `useData` hook.
        ssg: {
          stars: repo.stargazers_count
        }
      },
      // The page will be considered as stale and regenerated every 60 seconds.
      revalidate: 60
    }))
}
 
export const Stars = () => {
  // Get the data from SSG, and render it as a component.
  const { stars } = useData()
  return <strong>{stars}</strong>
}
 
Nextra has <Stars /> stars on GitHub!
```
 
## Method 2
https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props#using-getstaticprops-to-fetch-data-from-a-cms

I'm created a new page `text.tsx` in `/pages` folder with code:

```
export default function Blog({ data }) {
  return (
    <ul>
      {data.map((post) => (
        <li>{post.username}</li>
      ))}
    </ul>
  )
}

export async function getStaticProps() {
  const res = await fetch('http://localhost:3000/api/getSponsors')
  const data = await res.json()

  return {
    props: {
      data,
    },
  }
}
```

This is work, but page don't have UI elements (header, sidebar and etc):

<img width="175" alt="Screenshot" src="https://github.com/shuding/nextra/assets/153316370/40b66482-4098-4176-a127-73d27428e803">

P.S. method 2 not work in `production` mode, only in `dev`, get this error when build for production :(

```
Error occurred prerendering page "/sponsors". Read more: https://nextjs.org/docs/messages/prerender-error

TypeError: fetch failed
```

**Reply by caffeine-addictt** [reply]
For your first example, you need to call the JSX `<Stars />` for it to show up.

**Reply by mrvectopus** [reply]
Oh yeah, I'm forgot about it. Thanks!

**Result:**
<img width="500" alt="Screenshot" src="https://github.com/shuding/nextra/assets/153316370/11b65414-b256-47d2-b7c8-f94f8042056f">

For the test I'm try change `donate: data.donate` to `donate: data`:
```
ssg: {
        donate: data || null,
        username: data.username || null,
      },
```

and get this error:

```
Server Error
Error: Objects are not valid as a React child (found: object with keys {donate, group, group_slug, id, message, name, username, xata}). If you meant to render a collection of children, use an array instead.

This error happened while generating the page. Any console logs will be displayed in the terminal window.
```

**Reply by caffeine-addictt** [reply]
You cannot render the object itself in JSX.
You are essentially trying to do
```tsx
const data = { donate: val1, username: val2 }

<div>{data}</div>
```

**Reply by mrvectopus** [reply]
I'm try to show on page a data from my API route with server side generation :(

**Reply by caffeine-addictt** [reply]
Perhaps I was not clear enough.

You changed your server code from
```jsx
{
    ssg: {
        donate: data.donate
        username: data.username
      },
}
```
to
```jsx
{
    ssg: {
        donate: data
        username: data.username
      },
}
```

Now when you call `useData()`, donate now is an object and no longer a render-able type.
```tsx
const { username, donate } = useData()

username // Type: string
donate // Type: Object { username: string, donate: any }
```

So now when you try to render `donate`, you are trying to render an `Object` which is not supported by JSX.

```tsx
export const Stars = () => {
  const { donate, username } = useData() // donate now is equal to { username: '', donate: ? }
  return (
      <span className="text-blue-600">This is donate {donate}, from {username}</span>
  );
};
```

**Reply by mrvectopus** [reply]
I'm change `donate: data.donate` to `donate: data` only for the test...
I need show this data {username}, {donate} from my API router: 

`
[{"donate":"slowly reward","group":"temporise inside","group_slug":"out deceivingly","id":"rec_cneaqp2k1n2tltosuvo0","message":"angelic woefully","name":"inquisitively gah","username":"over childhood","xata":{"createdAt":"2024-02-26T15:25:24.894Z","updatedAt":"2024-02-26T15:25:24.894Z","version":0}},{"donate":"supposing whether","group":"ugh crab","group_slug":"aha clearly","id":"rec_cneaqp2k1n2tltosuvog","message":"upwardly uh-huh","name":"register limp","username":"quizzically stockpile","xata":{"createdAt":"2024-02-26T15:25:24.894Z","updatedAt":"2024-02-26T15:25:24.894Z","version":0}}] 
`

"donate":"slowly reward"
"donate":"supposing whether"
"username":"over childhood"
"username":"quizzically stockpile"

and I'm add `null` because without `null` get this error:

```
ssg: {
        donate: data || null,
        username: data.username || null,
      },
```
```
Server Error
Error: Error serializing `.ssg.donate` returned from `getStaticProps` in "/welcome".
Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value.
```

**Reply by caffeine-addictt** [reply, marked as answer]
Oh, I didn't catch the array.

It should be
```jsx
donate: data[0].donate ?? '',
username: data[0].donate ?? ''
```

**Reply by mrvectopus** [reply]
You are my hero!!!

<img width="600" alt="Screenshot" src="https://github.com/shuding/nextra/assets/153316370/19885f07-59fb-481e-9403-2737e4e5bf86">

![happy-cat-happy-happy-cat](https://github.com/shuding/nextra/assets/153316370/d5d8384d-9bba-4950-9bc6-da29a1e6c95c)
